File: POTD/April'24/Medium/Python Implementation/Number_of_Wonderful_Substrings.py
class Solution:
    def wonderfulSubstrings(self, word: str) -> int:
        mp = {0:1}
        ans = 0
        cumXor = 0
        for i in range(0,len(word),1):
            shift = ord(word[i]) - ord('a')  # ord() returns the ASCII code for a one-character string
            cumXor ^= (1 << shift)
            ans += mp.get(cumXor,0)
            for shift in range(0,11,1):
                ans += mp.get(cumXor ^ (1 << shift),0)
            mp[cumXor]  = mp.get(cumXor,0) + 1 # mp.get(cumXor,0) returns the value of cumXor if it is present in the dictionary, otherwise it returns 0 .This method is to avoid keyError.
        return ans 

File: POTD/April'24/Medium/Number_of_Wonderful_Substrings.cpp
class Solution {
public:
    long long wonderfulSubstrings(string word) {
        map<long long,long long> mp;
        mp[0] = 1; // For Empty String
        long long ans = 0,cumXor = 0;

        for(auto ch : word){
            long long shift = ch - 'a';
            cumXor ^= (1 << shift);
            ans += mp[cumXor]; // For even frequencies
            for(char c = 'a';c <= 'j';c++){ // For Odd(1)-Even frequencies
                long long shift = c - 'a';
                ans += mp[cumXor ^ (1 << shift)];
            }
            mp[cumXor]++;
        }
        return ans;
    }
};

